python/cuspatial/cuspatial/core/_column/geometa.py

```diff
@@ -9,7 +9,10 @@
 import cudf
 
 
+# This causes arrow to encode NONE as =255, which I'll accept now
+# in order to keep the rest of the enums the same.
 class Feature_Enum(Enum):
+    NONE = -1
```

Comment:
What is the difference between a NONE row and a null row? Could this just be implemented using nulls?

Reply:
Good idea I'm looking into this now.

Reply:
I think the short answer is
```
tests/test_empty.py:64: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
core/geoseries.py:353: in to_pandas
    return self.to_geopandas()
core/geoseries.py:343: in to_geopandas
    final_union_slice.to_shapely(),
core/geoseries.py:405: in to_shapely
    union = self.to_arrow()
core/geoseries.py:508: in to_arrow
    return pa.UnionArray.from_dense(
pyarrow/array.pxi:2308: in pyarrow.lib.UnionArray.from_dense
    ???
pyarrow/error.pxi:144: in pyarrow.lib.pyarrow_internal_check_status
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pyarrow.lib.ArrowInvalid: Union type ids may not have nulls

pyarrow/error.pxi:100: ArrowInvalid
```

null is not supported by `DenseUnion`, so I created an Enum type to store the null rows. The consquences are minor, the nulls are stored in the denseunion offsets and types and nowhere else.

Reply:
Are we going to need to deal with the None type in C++?

Reply:
Out of curiosity, where does this null row come from? According to [spec](https://arrow.apache.org/docs/format/Columnar.html#union-layout):

```
Unlike other data types, unions do not have their own validity bitmap. Instead, the nullness of each slot is determined exclusively by the child arrays which are composed to create the union.
```

Do we actually have an use case where the null row is caused not by a child row being null (like a null point), but actually representing a type-less null row?

Reply:
I tried setting nulls in the `union` values first, as you could see. I'll look at having valid `union` values but a null rows where it is referenced and share my work.

Reply:
It seems to be the latter case, a type-less null row. You can construct a test example like this:

```
polygons = gpd.GeoSeries([
        Polygon(((-8, -8), (-8, 8), (8, 8), (8, -8))),
        Polygon(((-2, -2), (-2, 2), (2, 2), (2, -2))),
    ])
polygon = gpd.GeoSeries(polygons[0]) 

print(polygon.contains(polygons))

print(polygon.align(polygons))
```
```
(0    POLYGON ((-8.00000 -8.00000, -8.00000 8.00000,...
1                                                 None
dtype: geometry, 0    POLYGON ((-8.00000 -8.00000, -8.00000 8.00000,...
1    POLYGON ((-2.00000 -2.00000, -2.00000 2.00000,...
dtype: geometry)
0     True
1    False
dtype: bool
```

Reply:
`.contains` takes an argument to align or not align the input series. If called with aligns=True, the default, then an empty row is created via `.align` for each row that doesn't exist in the other series. The missing row is not a none-polygon as far as I can tell, just a None row.

Reply:
It is less trivial to try to create a new value in the `DenseUnion` that is one of the four valid types, but is null itself, because a type needs to be chosen as the "null placeholder", then where in [geoseries](https://github.com/rapidsai/cuspatial/pull/760/files/9198f6c284e2797d7103782d71ad4c3661800d9a#diff-c522c9afb3364b1aed2b2589c0d0c260dbc634bc54844536b1d382cecb92bf29R531-R533) when None enums are created, they'll also need to be inserted into the child array, with new offsets generated and both arrays properly arranged to reflect that.

Reply:
It is not my intention that you will deal with the `None` type at all in C++, a null row to cuSpatial pandas will mean no-op.

Reply:
Thank you for detailing the example! I definitely didn't think of the case of a null row from `contains`. Your current proposal suggest using a type code of `-1` and set all offset buffers to be `-1` for the null rows. I think for `-1` offset may lead to OOB access if not careful. How about storing an additional null type that maps to [`pa.null()`](https://arrow.apache.org/docs/python/generated/pyarrow.null.html#pyarrow.null) and store an extra `pa.NullArray` of length 1 and always use offset 0 for all null rows? The benefit is that allows you to treat the null rows homogeneously as other types in gather/scatter operations without the need to handle specially. Just a proposal - not saying this is definitely better than yours solution.

Reply:
I've put a few hours into attempting this alternative. It segfaults often when I am trying to access the `DenseUnion` null array at position 0, I can't seem to keep a value inside of it. Not only is it difficult to develop for because of the segfaults but it is much more expansive of a modification, with 6+ files including the geopandas adapters needing various changes to add the null buffer and null values. I'm inclined to keep this refactoring as is as it has a relatively minor amount of code modifications. We can look at the alternative as it becomes necessary, if so.

Reply:
You can play with it [here](https://github.com/rapidsai/cuspatial/pull/780) today if you want, but I'd like to get this merged soon if not.